dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,277 @@
+from dvc.remote.base import RemoteBASE
+from dvc.config import Config
+from dvc.exceptions import DvcException
+from dvc.remote.gdrive.pydrive import (
+    RequestListFile,
+    RequestListFilePaginated,
+    RequestCreateFolder,
+    RequestUploadFile,
+    RequestDownloadFile,
+)
+from dvc.remote.gdrive.utils import FOLDER_MIME_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+    GDRIVE_USER_CREDENTIALS_DATA = "GDRIVE_USER_CREDENTIALS_DATA"
+    DEFAULT_USER_CREDENTIALS_FILE = ".dvc/tmp/gdrive-user-credentials.json"
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
+        self.gdrive_client_id = self.config.get(
+            Config.SECTION_GDRIVE_CLIENT_ID, None
+        )
+        self.gdrive_client_secret = self.config.get(
+            Config.SECTION_GDRIVE_CLIENT_SECRET, None
+        )
+        if not self.gdrive_client_id or not self.gdrive_client_secret:
+            raise DvcException(
+                "Please specify Google Drive's client id and "
+                "secret in DVC's config. Learn more at "
+                "https://man.dvc.org/remote/add."
+            )
+        self.gdrive_user_credentials_path = self.config.get(
+            Config.SECTION_GDRIVE_USER_CREDENTIALS_FILE,
+            self.DEFAULT_USER_CREDENTIALS_FILE,
+        )
+
+        self.root_id = self.get_path_id(self.path_info, create=True)
+        self.cached_dirs, self.cached_ids = self.cache_root_dirs()
+
+    @on_exception(expo, DvcException, max_tries=8)
+    def execute_request(self, request):
+        try:
+            result = request.execute()
+        except Exception as exception:
+            retry_codes = ["403", "500", "502", "503", "504"]
+            if any(code in str(exception) for code in retry_codes):
+                raise DvcException("Google API request failed")
+            raise
+        return result
+
+    def list_drive_item(self, query):
+        list_request = RequestListFilePaginated(self.drive, query)
+        page_list = self.execute_request(list_request)
+        while page_list:
+            for item in page_list:
+                yield item
+            page_list = self.execute_request(list_request)
+
+    def cache_root_dirs(self):
+        cached_dirs = {}
+        cached_ids = {}
+        for dir1 in self.list_drive_item(
+            "'{}' in parents and trashed=false".format(self.root_id)
+        ):
+            cached_dirs.setdefault(dir1["title"], []).append(dir1["id"])
```

Comment:
Do I understand correctly that single title might refer to several ids?

Reply:
Yes, `id` is unique field which is generated by remote or can be requested ahead to be used for new remote items (folder, files, etc) creation.
During parallel upload of dvc files few dirs can be created with the same `title`, but different `id`.

Reply:
DVC treats its remotes kind of filesystem, so this `title -> id` correspondence should be 1 to 1 as long as dvc constructs/uploads it. If that is not the case this is probably an issue (some race?) during upload. If that happens then we should probably place some guards to prevent it.

Reply:
This conversation moved into another discussion thread https://github.com/iterative/dvc/pull/2551#discussion_r347577585

Reply:
@MaxRis  Resolve this then?

Reply:
@Suor please resolve it if no more questions there :)